## Compare Multiplication in Numpy

The differences between np.dot(), np.matmul(), "*" and "@" are confusing to me. So I decide to experiment a bit to understand how it works.

Additional numpy matrix operation not covered here:
- np.inner
- np.outer 
- np.vdot
- np.tensordot

**[Reference:](https://docs.scipy.org/doc/numpy/reference/routines.linalg.html)**

Linear algebra (numpy.linalg)
- Matrix and vector products

Function | Description
--- | ---
dot(a, b[, out]) |	Dot product of two arrays.
linalg.multi_dot(arrays)	| Compute the dot product of two or more arrays in a single function call, while automatically selecting the fastest evaluation order.
vdot(a, b)	| Return the dot product of two vectors.
inner(a, b)	| Inner product of two arrays.
outer(a, b[, out])	| Compute the outer product of two vectors.
matmul(x1, x2, /[, out, casting, order, …])	| Matrix product of two arrays.
tensordot(a, b[, axes])	| Compute tensor dot product along specified axes for arrays >= 1-D.
einsum(subscripts, *operands[, out, dtype, …])	| Evaluates the Einstein summation convention on the operands.
einsum_path(subscripts, *operands[, optimize])	| Evaluates the lowest cost contraction order for an einsum expression by considering the creation of intermediate arrays.
linalg.matrix_power(a, n)	| Raise a square matrix to the (integer) power n.
kron(a, b)	| Kronecker product of two arrays.

In [0]:
import numpy as np

In [0]:
print(np.__version__)

1.16.4


### Reference

[**numpy.dot:**](https://docs.scipy.org/doc/numpy/reference/generated/numpy.dot.html?highlight=matrix%20multiplication)


Dot product of two arrays. Specifically,

* If both a and b are 1-D arrays, it is inner product of vectors (without complex conjugation).

* If both a and b are 2-D arrays, it is matrix multiplication, but using matmul or a @ b is preferred.

* If either a or b is 0-D (scalar), it is equivalent to multiply and using numpy.multiply(a, b) or a * b is preferred.

* If a is an N-D array and b is a 1-D array, it is a sum product over the last axis of a and b.

* If a is an N-D array and b is an M-D array (where M>=2), it is a sum product over the last axis of a and the second-to-last axis of b:

`
dot(a, b)[i,j,k,m] = sum(a[i,j,:] * b[k,:,m])
`

[**numpy.matmul**](https://docs.scipy.org/doc/numpy/reference/generated/numpy.tensordot.html)

numpy.matmul(x1, x2, /, out=None, *, casting='same_kind', order='K', dtype=None, subok=True[, signature, extobj]) = <ufunc 'matmul'>

- Matrix product of two arrays.

Parameters:	
x1, x2 : array_like
Input arrays, *scalars not allowed.*

out : ndarray, optional
A location into which the result is stored. If provided, it must have a shape that matches the signature (n,k),(k,m)->(n,m). If not provided or None, a freshly-allocated array is returned.

**kwargs
For other keyword-only arguments, see the ufunc docs.

..versionadded:: 1.16
Now handles ufunc kwargs

Returns:	
y : ndarray
The matrix product of the inputs. This is a scalar only when both x1, x2 are 1-d vectors.

Raises:	
ValueError
If the last dimension of a is not the same size as the second-to-last dimension of b.

If a scalar value is passed in.

The behavior depends on the arguments in the following way.

- If both arguments are 2-D they are multiplied like conventional matrices.
- If either argument is N-D, N > 2, it is treated as a stack of matrices residing in the last two indexes and broadcast accordingly.
- If the first argument is 1-D, it is promoted to a matrix by prepending a 1 to its dimensions. After matrix multiplication the prepended 1 is removed.
- If the second argument is 1-D, it is promoted to a matrix by appending a 1 to its dimensions. After matrix multiplication the appended 1 is removed.

[**"@"**
](https://docs.python.org/3/whatsnew/3.5.html#whatsnew-pep-465)

The @ operator calls the array's __matmul__ method, not dot. This method is also present in the API as the function np.matmul

[**numpy.multiply**](https://docs.scipy.org/doc/numpy/reference/generated/numpy.multiply.html)

Equivalent to x1 * x2 in terms of array broadcasting.

Multiply arguments **element-wise.**

### Dimensions Testing

#### Scalar

In [0]:
a = 1
b = 2

In [0]:
a * b

2

In [0]:
np.multiply(a, b)

2

In [0]:
np.dot(a,b)

2

In [0]:
a @ b

In [0]:
np.matmul(a,b)

**a @ b**
or 
**np.matmul(a,b)**
is not supported

matmul differs from dot in two important ways.

* Multiplication by scalars is not allowed.
* Stacks of matrices are broadcast together as if the matrices were elements.

What does it mean by broadcast together as if the matrices were elements?

[Reference](https://stackoverflow.com/questions/34142485/difference-between-numpy-dot-and-python-3-5-matrix-multiplication)

**dot** and **matmul** methods behave differently when passed 3D (or higher dimensional) arrays. Quoting from the documentation some more:

**matmul**:

If either argument is N-D, N > 2, it is treated as a stack of matrices residing in the last two indexes and broadcast accordingly.

**dot**: 

For 2-D arrays it is equivalent to matrix multiplication, and for 1-D arrays to inner product of vectors (without complex conjugation). For N dimensions it is a sum product over the last axis of a and the second-to-last of b

In [0]:
np.matmul([a],[b]) # input not allow scalar

2

In [0]:
np.array([a]) @ np.array([b]) # normal python list not supported by '@'

2

#### 1D array with scalar

In [0]:
a = [1, 2]
b = 3

In [2]:
a * b

[1, 2, 1, 2, 1, 2]

In [5]:
np.multiply(a, b)

array([3, 6])

-> List multiply with n is copy and extend list itself n times

-> **np.dot** is element-wise, which is equal to numpy array multiplication

In [0]:
np.dot(a, b)

array([3, 6])

In [0]:
np.array(a) * b # same as np.multiply(a, b), because of array broadcasting

array([3, 6])

**Ref.** [numpy broadcasting](https://docs.scipy.org/doc/numpy-1.15.0/user/basics.broadcasting.html)

[Ref. 2](https://scipy-lectures.org/intro/numpy/operations.html#broadcasting)

In [0]:
np.multiply(a, b)

array([3, 6])

**np.matmul** only accepts ndarray

In [6]:
np.array(a).shape

(2,)

In [0]:
np.array(a)[:,None].shape

(2, 1)

In [7]:
np.array(a)[:,None]

array([[1],
       [2]])

In [0]:
np.matmul(np.array(a)[:,None], [b])

array([3, 6])

#### 1D array & 1D array

In [0]:
a = [1,2]
b = [2,3]
c = [3,4,5]

In [9]:
a * b

TypeError: ignored

In [0]:
np.multiply(a,b) # treat as numpy array element-wise, (a,c) won't work becasue of shape difference

array([2, 6])

In [0]:
np.asarray(a) * np.asarray(b)

array([2, 6])

In [10]:
np.array(a) * b

array([2, 6])

-> After convert to numpy array, two 1D arrays multiplication is element-wise using operation "*"

Ref. [np.array vs np.asarray](https://stackoverflow.com/questions/14415741/numpy-array-vs-asarray)

**np.dot**, **np.matmul** and **@** give same result for 1D array x 1D array, return dot product of 2 array

- Use '@' after confirm it is operating on numpy object

In [0]:
np.dot(a,b) # 1x2 + 2x3

8

In [0]:
np.matmul(a,b) # same matrix multiplication as np.dot

8

In [0]:
np.array(a) @ np.asarray(b) # convert to numpy array before using `@`

8

Generally it is a good practice to convert to numpy array first, instead of using python's default list, unless need some special functionality

#### 2D array & 1D array

In [0]:
a = np.array([[1,2],
              [3,4]])
b = np.array([5,6])

In [0]:
a * b # same as np.multiply(a,b), element-wise product (hadamard product)

array([[ 5, 12],
       [15, 24]])

In [0]:
a @ b # same as np.matmul(a,b) and np.dot(a,b), dot product

array([17, 39])

In [0]:
a.dot(b)
# [5 + 12, 15 + 24]

array([17, 39])

In [0]:
b.dot(a)
# [5 + 18, 10 + 24]

array([23, 34])

In [0]:
b @ a

array([23, 34])

#### 2D array & 2D array

In [0]:
a = np.array([[1,2],
              [3,4]])
b = np.array([[5,6],
              [7,8]])

In [0]:
a * b # same as np.multiply(a,b) -> a.shape == b.shape (because of element-wise)

array([[ 5, 12],
       [21, 32]])

In [0]:
a @ b # same as np.matmul(a,b) -> a.shape[1] == b.shape[0] (because of matrix multiplication)

array([[19, 22],
       [43, 50]])

In [0]:
np.dot(a,b) # same dot product as "a @ b"

array([[19, 22],
       [43, 50]])

If both a and b are 2-D arrays, it is matrix multiplication, but using matmul or a @ b is preferred.
[Ref. ](https://stackoverflow.com/questions/52062496/why-is-a-dotb-faster-than-ab-although-numpy-recommends-ab)

**Exception** *np.matrix class*

a * b is no longer element-wise but matrix multiplication

In [0]:
a = np.matrix([[1,2],
               [3,4]])
b = np.matrix([[5,6],
               [7,8]])

In [0]:
a * b

matrix([[19, 22],
        [43, 50]])

In [0]:
a @ b

matrix([[19, 22],
        [43, 50]])

In [0]:
a.dot(b)

matrix([[19, 22],
        [43, 50]])

In [14]:
b @ a

matrix([[23, 34],
        [31, 46]])

"*", "@" and np.dot give same result

#### 3D array & 1D array

In [0]:
a = np.array([[[1,2,3],
               [2,3,4]],
              [[3,4,5],
               [4,5,6]]])
b = np.array([1,2,3])

In [0]:
print(a.shape)
print(b.shape)

(2, 2, 3)
(3,)


In [0]:
a * b # same as np.multiply(a,b), broadcasting
# b -> [[[1,2,3],
#        [1,2,3]],
#       [[1,2,3],
#        [1,2,3]]]

array([[[ 1,  4,  9],
        [ 2,  6, 12]],

       [[ 3,  8, 15],
        [ 4, 10, 18]]])

**np.dot()**

- For 2-D arrays it is equivalent to matrix multiplication
- For 1-D arrays to inner product of vectors (without complex conjugation).
- For N dimensions it is a sum product over the last axis of a and the second-to-last of b

In [0]:
a.dot(b)
# [[1+4+ 9], [2+ 6+12],
#  [3+8+15], [4+10+18]]

array([[14, 20],
       [26, 32]])

**np.matmul**

If either argument is N-D, N > 2, it is treated as a stack of matrices residing in the last two indexes and broadcast accordingly.

In [0]:
np.matmul(a,b) # a @ b

array([[14, 20],
       [26, 32]])

##### Reverse - 1D vs 3D

In [0]:
b * a # same result

array([[[ 1,  4,  9],
        [ 2,  6, 12]],

       [[ 3,  8, 15],
        [ 4, 10, 18]]])

In [0]:
c = np.array([1,2]) # need to match dim for dot product

In [0]:
c.dot(a)

array([[ 5,  8, 11],
       [11, 14, 17]])

In [0]:
c @ a

array([[ 5,  8, 11],
       [11, 14, 17]])

#### 3D array & 2D array

In [15]:
a = np.arange(12).reshape((2, 2, 3))
b = np.arange(6).reshape((3,2))
print(a)
print(b)
print(a.shape)
print(b.shape)

[[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]]
[[0 1]
 [2 3]
 [4 5]]
(2, 2, 3)
(3, 2)


In [16]:
a * b

ValueError: ignored

In [0]:
a * b.T # np.multiply(a,b), broadcast and element-wise, shape of b can be (3,) or (2,3) or (2,2,3)

array([[[ 0,  2,  8],
        [ 3, 12, 25]],

       [[ 0, 14, 32],
        [ 9, 30, 55]]])

In [0]:
a.dot(b)
# 2x2x3 x 3x2 = 2x2x2
# Result:
# [[[0+ 2+ 8, 0+ 3+10],
#   [0+ 8+20, 3+12+25]],
#
#  [[0+14+32, 6+21+40],
#   [0+20+44, 9+30+55]]]

array([[[10, 13],
        [28, 40]],

       [[46, 67],
        [64, 94]]])

In [0]:
a @ b # same as dot product

array([[[10, 13],
        [28, 40]],

       [[46, 67],
        [64, 94]]])

##### Reverse - 2D & 3D

In [0]:
a = np.arange(12).reshape((2, 2, 3))
b = np.arange(6).reshape((3,2))
print(a)
print(b)
print(a.shape)
print(b.shape)

[[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]]
[[0 1]
 [2 3]
 [4 5]]
(2, 2, 3)
(3, 2)


In [0]:
b.T * a # same result

array([[[ 0,  2,  8],
        [ 3, 12, 25]],

       [[ 0, 14, 32],
        [ 9, 30, 55]]])

In [0]:
b.dot(a)

array([[[ 3,  4,  5],
        [ 9, 10, 11]],

       [[ 9, 14, 19],
        [39, 44, 49]],

       [[15, 24, 33],
        [69, 78, 87]]])

In [0]:
b @ a

array([[[ 3,  4,  5],
        [ 9, 14, 19],
        [15, 24, 33]],

       [[ 9, 10, 11],
        [39, 44, 49],
        [69, 78, 87]]])

np.dot and np.matmul behavior differently when it comes to ND array

matmul -> Stacks of matrices are broadcast together as if the matrices were elements.

It only treats last 2 columns as matrix and do 2D matrix multiplication

dot -> it is a sum product over the last axis of a and the second-to-last of b

##### Anther example

In [0]:
a = np.random.rand(2,2,2)
b = np.random.rand(2,2)

for 3D x 2D, the result is the same

In [0]:
a @ b

array([[[0.56374055, 0.08640099],
        [0.7334552 , 0.10388443]],

       [[1.41197078, 0.19617761],
        [0.4764388 , 0.06455949]]])

In [0]:
a.dot(b)

array([[[0.56374055, 0.08640099],
        [0.7334552 , 0.10388443]],

       [[1.41197078, 0.19617761],
        [0.4764388 , 0.06455949]]])

for 2D x 3D , the result is different

In [0]:
b @ a

array([[[0.06913136, 0.74882671],
        [0.06167257, 0.59849362]],

       [[0.83013132, 0.81854009],
        [0.65837523, 0.64639895]]])

In [0]:
b.dot(a)

array([[[0.06913136, 0.74882671],
        [0.83013132, 0.81854009]],

       [[0.06167257, 0.59849362],
        [0.65837523, 0.64639895]]])